### DISCUSSION TITLE
How are multiple clients handled?

### DISCUSSION THREAD

**Discussion by Super-Beast** [LizardByte/community#258] [Sunshine]
I'm interested in setting up a server using Sunshine. I'd like to allow multiple users to play their own games individually. Is this possible or do all clients see the exact same screen as what's displayed on the host? Say one user wants to play Terraria on their PC and another user wants to play Minecraft on their android phone. Is this possible without the inputs of each mixing up with eachother and without the users seeing the other users screen?

**Comment by ReenigneArcher** [comment]
All clients view the same screen and can play the same game, if the game supports multiplayer.

The use case you're interested in, is not possible on a single OS. You will need to either create virtual machines, each with their own GPU, display, etc... or have entirely different computers.

This is a limitation of the OSes themselves. Every desktop environment I've ever used only allows you to control a single active application. The other applications are not active and do not receive input (keyboard, mouse, gamepad).

**Reply by davispuh** [reply]
> The use case you're interested in, is not possible on a single OS.

> This is a limitation of the OSes themselves. Every desktop environment I've ever used only allows you to control a single active application. The other applications are not active and do not receive input (keyboard, mouse, gamepad).

This is not true at all. On Windows you can patch `termsrv.dll` to allow multiple RDP sessions. See https://woshub.com/how-to-allow-multiple-rdp-sessions-in-windows-10/

On Linux you also can run multiple RDP/VNC sessions.

**Reply by ReenigneArcher** [reply]
Obviously I was talking about the case of a single session, which is where Sunshine operates. We don't use RDP sessions.

**Reply by davispuh** [reply]
Well this is a feature request for this feature and I think it should be doable to implement :) Not necessarily using RDP (maybe only way?) but whichever way to accomplish this goal.

**Comment by igorbienkowski** [comment]
Well this is very interesting case and i believe it should be possible to implement such function in the host. It might work similar to OBS where we could set the "window" that we want to stream and considering that there ate many game "bots" that are handling input independently from users mouse and keyboard. If this two features would be combined, i believe it would be possible to play even on 3 clients from one host. Of course considering that the game would be fe. terraria (which doesn't require too much resources) and the host machine would be powerfull enough to run 3 instances of the game at once. This would be really big task to implement and all what i've mentioned above is just "maybe". Considering that Sunshine is open source and noone is earning any money on it i believe it would take a lot of time, effort and nerves to implement something like that :)

**Reply by ReenigneArcher** [reply]
You'd only be able to control a single instance of the game at once, so it's kind of a moot point.

**Reply by brokensyntax** [reply]
Linux being a mutli-user OS by design allows several users to interact with it remotely, each with their own sessions. Whether TTY or X sessions.

For the sake of memory management and process namespaces, it might be easier to roll a solution using containers where each session drops into its own container, maybe this will be my intro to k8s project? :D

The main thing is that each client will need to connect with its own session. (Think like when you SSH to a linux system with X11 forwarding enabled. Everyone on that system can spawn their own X programs and their output and controls are bound to that session, no user conflicts.)

**Comment by davispuh** [comment]
I am also interested in this for Linux.

I don't really know how RDP/VNC works but we need something like that here and with GPU acceleration.

I tried running multiple Sunshine instances under different users but they all shared same "desktop", maybe there is some configuration to achieve this?

Maybe this could be implemented with some help from Systemd, KWin, [Venus](https://docs.mesa3d.org/drivers/venus.html)?

**Reply by brokensyntax** [reply]
From the MS RDP side of things, you need the RDP connection drivers, and server executable that comes with Windows Server (the related DLLs can be migrated to a non-server system, but licensing TOS etc.).   
   
Remote Desktop Server allows for multiple user sessions and divies available resources according to demand, or configuration. The difficulty of these configurations scales up based on design and use-case. (MS RD Gateway, MS RD Connection Brokers, MS RD Farms, etc.) and tend to be pretty picky about trusting the certificate presented by the server; so self-signed is a lot of manual annoyance.   
   
   Without a server-level Windows operating system, it isn't worth chasing RDP back-bones, but newer releases of Windows and Windows server do natively support SSH, and with WSLg, some level of graphics/x11 forwarding.    
       
   While I've used SPICE, and like its performance, I haven't used it extensively enough to provide any insights about the potential of that as the presentation layer.

**Comment by GiladDrori** [comment]
I would also love this as a feature. Getting rid of the overhead of a whole VM just for a parallel session would be a game changer.

**Reply by brokensyntax** [reply]
Oof yeah, VMs going to be inefficient, done any testing with docker? Can we containerize each application into its own presentation host?

**Comment by gschintgen** [comment]
Hi all, there is this little project: https://github.com/games-on-whales/wolf
"Wolf is a streaming server for [Moonlight](https://moonlight-stream.org/) that allows you to share a single server with multiple remote clients in order to play videogames!"

I haven't tried it out myself, but it might suit your needs.
